ansible_sdk/_util/metrics.py

```diff
@@ -0,0 +1,122 @@
+        module_csv_writer = CSVAsync(module_csv_fh, headers, restval="NULL")
+        await module_csv_writer.writeheader_async()
+
+        async for ev in status_obj.events:
+            end_time = datetime.strptime(ev.get("created"), "%Y-%m-%dT%H:%M:%S.%f")
+            if ev.get("event") == "playbook_on_start":
+                metrics_data["started"] = datetime.strptime(
+                    ev.get("created"), "%Y-%m-%dT%H:%M:%S.%f"
+                )
+            if ev.get("event") == "playbook_on_task_start":
+                metrics_data["task_counter"] += 1
+                # Update the count of task
+                resolved_task_name = ev["event_data"]["resolved_action"]
+                if resolved_task_name in metrics_data["task_data"]:
+                    metrics_data["task_data"][resolved_task_name] += 1
+                else:
+                    metrics_data["task_data"][resolved_task_name] = 1
+
+                # Update the count of role
+                resolved_role_name = ev["event_data"].get("role")
+                if resolved_role_name:
+                    if resolved_role_name in metrics_data["role_data"]:
+                        metrics_data["role_data"][resolved_role_name] += 1
+                    else:
+                        metrics_data["role_data"][resolved_role_name] = 1
+
+            if ev.get("event") == "playbook_on_stats":
+                # Write module metrics data
+                for task in metrics_data["task_data"]:
+                    await module_csv_writer.writerow_async(
+                        {
+                            "job_id": ev["runner_ident"],
+                            "module_fqcn": task,
+                            "role_fqcn": "",
+                            "task_count": metrics_data["task_data"][task],
+                            "task_duration": "",
+                        }
+                    )
+
+                # Write job metrics data
+                await job_csv_writer.writerow_async(
+                    {
+                        "job_id": ev["runner_ident"],
+                        "job_type": "local",
+                        "started": metrics_data.get("started"),
+                        "finished": end_time,
+                        "job_state": "",
+                        "hosts_ok": len(ev["event_data"]["ok"]),
+                        "hosts_changed": len(ev["event_data"]["changed"]),
+                        "hosts_skipped": len(ev["event_data"]["skipped"]),
+                        "hosts_failed": len(ev["event_data"]["failures"]),
+                        "hosts_unreachable": len(ev["event_data"]["failures"]),
+                        "task_count": metrics_data["task_counter"],
+                        "task_duration": end_time - metrics_data.get("started"),
+                    }
+                )
+
+            status_obj.drop_event(ev)
+
+        module_csv_fh.close()
```

Comment:
these two lines are unnecessary if the file access is done under a context manager